corehq/apps/api/tests/test_auth.py

```diff
@@ -0,0 +1,70 @@
+from django.contrib.auth.models import AnonymousUser
+from django.http import HttpResponse
+from django.test import TestCase, RequestFactory
+
+from corehq.apps.api.resources.auth import LoginAuthentication, LoginAndDomainAuthentication
+from corehq.apps.domain.models import Domain
+from corehq.apps.users.models import WebUser, HQApiKey
+
+
+class AuthenticationTestBase(TestCase):
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        cls.factory = RequestFactory()
+        cls.domain = Domain.get_or_create_with_name('api-test', is_active=True)
+        cls.username = 'alice@example.com'
+        cls.password = '***'
+        cls.user = WebUser.create(cls.domain.name, cls.username, cls.password, None, None)
+        cls.api_key, _ = HQApiKey.objects.get_or_create(user=WebUser.get_django_user(cls.user))
+
+    @classmethod
+    def tearDownClass(cls):
+        cls.domain.delete()
+        super().tearDownClass()
+
+    def _get_request_with_api_key(self):
+        return self._get_request(HTTP_AUTHORIZATION=f'ApiKey {self.username}:{self.api_key.key}')
+
+    def _get_request(self, domain=None, **extras):
+        path = self._get_domain_path() if domain else ''
+        request = self.factory.get(path, **extras)
+        request.user = AnonymousUser()  # this is required for HQ's permission classes to resolve
+        request.domain = domain  # as is this for any domain-specific request
+        return request
+
+    def _get_domain_path(self):
+        return f'/a/{self.domain.name}/'
+
+    def assertAuthenticationSuccess(self, auth_instance, request):
+        self.assertTrue(auth_instance.is_authenticated(request))
+
+    def assertAuthenticationFail(self, auth_instance, request):
+        result = auth_instance.is_authenticated(request)
+        # currently auth classes return a 401/403 response in some scenarios
+        # this should likely be changed to always return False
+        # more discussion here: https://github.com/dimagi/commcare-hq/pull/28201#discussion_r461082885
+        if isinstance(result, HttpResponse):
+            self.assertTrue(result.status_code in (401, 403))
+        else:
+            self.assertFalse(result)
+
+
+class LoginAuthenticationTest(AuthenticationTestBase):
+
+    def test_login_no_auth(self):
+        self.assertAuthenticationFail(LoginAuthentication(), self._get_request())
+
+    def test_login_with_auth(self):
+        self.assertAuthenticationSuccess(LoginAuthentication(), self._get_request_with_api_key())
+
+
+class LoginAndDomainAuthenticationTest(AuthenticationTestBase):
+
+    def test_login_with_domain(self):
+        self.assertAuthenticationSuccess(LoginAndDomainAuthentication(), self._get_request(domain=self.domain))
```

Comment:
shouldn't this fail since it has no auth?

Reply:
eagle eyes. Thanks, it was returning a 401 `HttpResponse` which was not failing on `assertTrue`. This whole "sometimes it's true or false and sometimes it's an `HttpResponse` thing is really bugging me. I'm pretty sure it's entirely unneccessary, but just haven't gone through enough code to change it yet. Hopefully soon. :pray: 

(fixed in the latest 3 commits, which revealed some other issues with the domain checks...)

Reply:
Seems like this is by design in Tastypie. Here's part of the docstring from `is_authenticated`:
```
Should return either ``True`` if allowed, ``False`` if not or an
        ``HttpResponse`` if you need something custom.
```

I guess the question is whether we should try and be more consistent and always return an HTTP response. I do think the custom messages are useful at times, particularly the one you linked.

Reply:
Yeah agreed, now that I get the use case I think the tradeoff between improvement in API docs is worth the additional code complexity